src/goto-programs/json_goto_trace.cpp

```diff
@@ -13,12 +13,95 @@ Author: Daniel Kroening
 
 #include "json_goto_trace.h"
 
-#include <cassert>
-
+#include <util/arith_tools.h>
+#include <util/config.h>
+#include <util/invariant.h>
 #include <util/json_expr.h>
+#include <util/simplify_expr.h>
 
 #include <langapi/language_util.h>
+#include <solvers/flattening/pointer_logic.h>
+
+/// Replaces in src, expressions of the form pointer_offset(constant) by that
+/// constant.
+/// \param src: an expression
+void remove_pointer_offsets(exprt &src)
+{
+  if(src.id()==ID_pointer_offset &&
+     src.op0().id()==ID_constant &&
+     src.op0().type().id()==ID_pointer)
+  {
+    std::string binary_str=id2string(to_constant_expr(src.op0()).get_value());
+    // The constant address consists of OBJECT-ID || OFFSET.
+    // Shift out the object-identifier bits, leaving only the offset:
+    mp_integer offset=binary2integer(binary_str.substr(BV_ADDR_BITS), false);
+    src=from_integer(offset, src.type());
+  }
+  else
+    for(auto &op : src.operands())
+      remove_pointer_offsets(op);
+}
 
+/// Replaces in src, expressions of the form pointer_offset(array_symbol) by a
+/// constant value of 0. This is meant to simplify array expressions.
+/// \param src: an expression
+/// \param array_symbol: a symbol expression representing an array
+void remove_pointer_offsets(exprt &src, const symbol_exprt &array_symbol)
+{
+  if(src.id()==ID_pointer_offset &&
+     src.op0().id()==ID_constant &&
+     src.op0().op0().id()==ID_address_of &&
+     src.op0().op0().op0().id()==ID_index)
```

Comment:
@romainbrenguier Would you happen to know where you had seen such expressions? (This originally was f4fa77802f43a7b.) As far as I am aware, an expression with id `ID_constant` should not have any operands.

Reply:
> One thing that confused me a bit is that expressions like `&a[0]` gets simplified by `from_expr` to `a`, but `simplify_expr` does not simplify it.

That's because as much as they can be used interchangeably at source level, their types are different: with `int a[1]`, `a` is of type `int[1]`, while `&a[0]` is of type `int*`. Consequently the pretty printer can do this "simplification" at text level, but transforming expressions would not be valid.

Reply:
ok thanks, it makes more sense to me now